src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,569 @@
+                var nodeStack = new Stack<int>();
+
+                void unblock(int node)
+                {
+                    if (blockedSet.Remove(node) && blockedMap.TryGetValue(node, out var nodesToUnblock))
+                    {
+                        blockedMap.Remove(node);
+                        foreach (var n in nodesToUnblock)
+                        {
+                            unblock(n);
+                        }
+                    }
+                }
+
+                bool populateCycles(int currNode)
+                {
+                    var foundCycle = false;
+                    nodeStack.Push(currNode);
+                    blockedSet.Add(currNode);
+
+                    foreach (var child in intputSCC[currNode])
+                    {
+                        if (filter.Contains(child))
+                        {
+                            if (child == startNode)
+                            {
+                                foundCycle = true;
+                                cycles.Add(nodeStack.Reverse().ToList());
+                            }
+                            else if (!blockedSet.Contains(child))
+                            {
+                                foundCycle |= populateCycles(child);
+                            }
+                        }
+                    }
+
+                    nodeStack.Pop();
+
+                    if (foundCycle)
+                    {
+                        unblock(currNode);
+                    }
+                    else
+                    {
+                        // Mark currNode as being blocked on each of its children
+                        // If any of currNode's children unblock, currNode will unblock
+                        foreach (var child in intputSCC[currNode])
+                        {
+                            if (filter.Contains(child))
+                            {
+                                if (!blockedMap.ContainsKey(child))
+                                {
+                                    blockedMap[child] = new HashSet<int>() { currNode };
+                                }
+                                else
+                                {
+                                    blockedMap[child].Add(currNode);
+                                }
+                            }
+                        }
+                    }
+
+                    return foundCycle;
+                }
+
+                populateCycles(startNode);
+
+                return cycles;
+            }
+        }
+
+        /// <summary>
+        /// This is a dictionary mapping source nodes to information about target nodes. This information is represented
+        /// by a dictionary mapping target node to the edges pointing from the source node to the target node.
+        /// </summary>
+        private readonly Dictionary<CallGraphNode, Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>> _Dependencies =
```

Comment:
It looks like you want to give the promise that all entries in this dictionary have been stripped of any positional information. Is that correct? If so, you should document it.

Reply:
It is really more that CallGraphEdge's ParamResolutions should not have any positional information. I have added that to its summary.